### PR TITLE
added evaluation code for T5

### PR DESCRIPTION
**Pull request by Robinysh**

### CODE REVIEW

File: evaluation/LM_baseline.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        break
+
+    return math.exp(total_loss / total_tokens)
+
+def generate_sentenceids(model, tokenizer, sentences):
+    sentenceids = []
+    for sent in sentences:
+        inputs = tokenizer(sent, return_tensors="pt")
+        input_ids = inputs["input_ids"].squeeze(0)
+        sentenceids.extend([input_ids[:i+1].unsqueeze(0) for i in range(len(input_ids))])
```
Comment (Robinysh) [inline evaluation/LM_baseline.py]: I dont understand the purpose of this part. Why are you creating each sub-sequence of sentence from start to i+1? I would expect you to use the tokenizer input_ids directly.

File: evaluation/LM_baseline.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    return sentenceids
+
+def calculate_throughput(model, tokenizer, sentences, verbose=False):
+    sentenceids = generate_sentenceids(model, tokenizer, sentences)
+    start_time = time.time()
+    for input_ids in sentenceids:
+        with torch.no_grad():
+            outputs = model.generate(input_ids, max_new_tokens=len(input_ids[0])+1)
```
Comment (Robinysh) [inline evaluation/LM_baseline.py]: model.generate will generate sentences. We are not generating sentences or sampling, but only calculating the logits of our given tokens. Simply use `model(input_ids)`.

File: evaluation/LM_baseline.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        
+        # print(tokenizer.decode(input_ids[0]))
+        # print(tokenizer.decode(outputs[0],
+        #                 skip_special_tokens = False))
+        # print('\n')
+    
+    end_time = time.time()
+
+    # calculate throughput using time difference
+    return len(sentences)/(end_time - start_time)
```
Comment (Robinysh) [inline evaluation/LM_baseline.py]: You should calculate token/s, not sentence/s. Sentence length can vary a lot depending on the dataset.

File: evaluation/LM_baseline.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        with torch.no_grad():
+            ids = input_ids[:i+1].unsqueeze(0)
+            sent = tokenizer.decode(ids[0])
+            # print(ids)
+            
+            outputs = model(input_ids = ids, labels = ids)
+            # print(outputs)
+        
+        # predictions = outputs.logits[:, -1, :]
+        loss = outputs.loss
```
Comment (Robinysh) [inline evaluation/LM_baseline.py]: I think the loss might already be normalized by sentence length, so the normalization for perplexity calculation is repeated. You may need to double check.

File: evaluation/LM_baseline.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+def calculate_perplexity_ex(model, tokenizer, input_text, top_k=5):
+    inputs = tokenizer(input_text, return_tensors="pt")
+
+    input_ids = inputs["input_ids"].squeeze(0)
+    # print(input_ids)
+
+    total_loss = 0
+    total_tokens = 0
+
+    # Loop through all position
```
Comment (Robinysh) [inline evaluation/LM_baseline.py]: I dont understand the purpose of this loop. You should be able to calculate the perplexity in one go by feeding in the entire sentence to the model.

File: evaluation/LM_baseline.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
+import torch
+import math
+import time
+
+def predict_next_token(logits, top_k, tokenizer):
```
Comment (Robinysh) [inline evaluation/LM_baseline.py]: I dont think you need a sampling function.

**Comment by Robinysh** [comment]
Put your code into `src/`, maybe `src/evalutation/`

**Review by Robinysh — COMMENTED** [review]
Put your code into `src/`, maybe `src/evalutation/`
